Support for for-loops

thisac opened the pull request:
**Context**: Currently Blackbird doesn't support for-loops and thus cannot iterate through lists of statements. This would be useful when having many similar statements.

**Description**: Support for parsing for-loops are added to Blackbird. The user can write e.g.:

```python
for int m in [2, 7, 3]
    MeasureFock() | m
```

to apply the operation to modes 2, 7 and 3, or

```python
par array phases =
    {phase_0}
    {phase_1}
    {phase_2}
    {phase_3}
    {phase_4}

for int m in 0:5:1
    MZgate(phases[m], phases[m+1]) | [m, m+1]
```

to apply the `MZgate` with indexing from the `phases` list for the range `0:5:1 = [0, 1, 2, 3, 4]`

Other additions:

* Indexing is now supported inside both modes and arguments (e.g.  `MZgate(list[2], list[0]) | [list[3],list[1]])`)
* Expressions are supported inside all indexes (e.g. `phases[-3 + 2*2]`).

josh146 commented on blackbird_python/blackbird/listener.py:
```diff
@@ -135,6 +135,7 @@ def __init__(self, cwd=None):
         self._program = BlackbirdProgram()
         self._includes = {}
         self._cwd = cwd
+        self._infor = False
```
What does this variable represent?

josh146 commented on blackbird_python/blackbird/listener.py:
```diff
@@ -341,12 +342,30 @@ def exitStatement(self, ctx: blackbirdParser.StatementContext):
             ctx: statement context
         """
+        if isinstance(ctx.parentCtx, blackbirdParser.ForloopContext):
+            if self._infor:
+                return
```
Might be worth commenting to show what is happening here 🤔

josh146 commented on blackbird_python/blackbird/listener.py:
```diff
@@ -341,12 +342,30 @@ def exitStatement(self, ctx: blackbirdParser.StatementContext):
             op = ctx.measure().getText()
 
-        modes = [int(i) for i in ctx.modes().getText().split(",")]
+        # get modes; if variables, replace with corresponding values
+        modes = ctx.arrayrow().getText().split(",")
+        for i, m in enumerate(modes):
+            try:
+                modes[i] = int(m)
+            except ValueError:
+                m = m.strip()
+                if m in _VAR.keys():
+                    if isinstance(_VAR[m], int):
+                        modes[i] = _VAR[m]
+                    else:
+                        raise ValueError(f"Mode must be of type int, not {type(_VAR[m])}")
+                else:
+                    raise ValueError(f"Variable {m} not declared")
```
This is really nice!

josh146 commented on blackbird_python/blackbird/listener.py:
```diff
@@ -427,6 +446,28 @@ def exitStatement(self, ctx: blackbirdParser.StatementContext):
         else:
             self._program._operations.append(operation)
 
+    def enterForloop(self, ctx: blackbirdParser.ForloopContext):
+        self._infor = True
+
+    def exitForloop(self, ctx: blackbirdParser.ForloopContext):
+        self._infor = False
```
Ah, it stands for 'in for loop' 😆 I would suggest changing it to `self._in_for`

josh146 replied:
I kept reading it as 'info-r'

thisac replied:
Yep, I completely agree. It was just a quickly named variable for testing at first, and then I just forgot to update it to something more legible. :laughing:

josh146 commented on blackbird_python/blackbird/listener.py:
```diff
@@ -427,6 +446,28 @@ def exitStatement(self, ctx: blackbirdParser.StatementContext):
+
+        if ctx.INT():
+            _FORVAR = range(int(ctx.INT().getText()))
+        elif ctx.vallist():
+            _FORVAR = ctx.vallist().getText().split(',')
```
Does the `vallist()` context assume the syntax `[1, 2, 3]` (e.g., do you have to use square brackets?)

thisac replied:
`vallist` is currently only the content of the list, i.e. `1, 2, 3`, so any brackets could in practice be used, although the for-loop only supports lists with square brackets.

josh146 commented on blackbird_python/blackbird/listener.py:
```diff
@@ -427,6 +446,28 @@ def exitStatement(self, ctx: blackbirdParser.StatementContext):
+            _FORVAR = ctx.vallist().getText().split(',')
+
+        for var in _FORVAR:
+            if ctx.NAME():
+                try:
+                    var = PYTHON_TYPES[ctx.vartype().getText()](var)
+                except ValueError:
+                    raise ValueError(f"invalid value {var}; must be {ctx.vartype().getText()}")
+                _VAR[ctx.NAME().getText()] = var
```
Could this be replaced with the function for evaluating an expression? That way you could have

```python
for [x**2, 54, 1] -> m
```

and it should still work.

I think the function for evaluating expressions also includes the logic for getting variable values.

thisac replied:
Done! Expressions are now supported. :slightly_smiling_face:

thisac commented:
> Does for 3 iterate over [0, 1, 2] like Python, or [0, 1, 2, 3]?

It should iterate over [0, 1, 2], since it creates a `range(3)` sequence. 

> Can you do for 3 -> m, to have m take the values 0, 1, 2 for each loop?

Yes, that's how it currently works.

> It would be good to extend this to support not just passing the number of iterations, but providing the start, end, and step values. E.g., for 6, 20, 2 -> m

Just implemented it with the syntax `for 6:20:2 -> m`. What do you think?

I updated the for-loop to support the same type of lists as the modes: `[1, 2, 3]`, `(1, 2, 3)` or `1, 2, 3`, and I think that the last option, without any brackets/parentheses might work best with the rest of the syntax (although that's debatable, and if you prefer to fix it to `[]` I'll do that). I also added support for lists that contain both parameters (e.g. `{phase_0}` or other types), and they are callable from within any statement, including for-loops. So, this should now work (although I haven't written any thorough tests yet):

```
name test_name
version 1.0
target fock (num_subsystems=1, cutoff_dim=7, shots=10)

par array phases =
    {phase_0}
    {phase_1}
    {phase_2}
    {phase_3}
    {phase_4}

float alpha = 0.3423
Coherent(alpha, sqrt(pi)) | 0
for 0, 3, 1 ->  int m
    MZgate(phases[m], phases[m+1]) | [m, m+1]

for 2:20:2 ->  int m
    MeasureFock() | m
```

Also, any type of "matrix" will currently be considered a list, by flattening it before attempting to extract a value from it. So this should work in the context above as well:

```
par array phases =
    {phase_0}, {phase_1}
    {phase_2}, {phase_3}
    {phase_4}
```

co9olguy commented:
Just curious about the philosophy for the chosen semantics here, as I found both
`for 3`
and 
`for [2, 7, 3] ->  int m` 
to be unintuitive

Is this something that is forced by our choice of lexer/parser?

Historical note: originally, blackbird was meant to be kept as close to "python" as possible, since the way a user declares their program in SF is essentially also "blackbird code" (with some small, unavoidable, modifications between "standalone blackbird" and "python-embedded blackbird"). Why would we want to do that here, where---at least to my naive eye---it doesn't seem necessary to depart from python-style?

thisac commented:
> Just curious about the philosophy for the chosen semantics here, as I found both
> `for 3`
> and
> `for [2, 7, 3] -> int m`
> to be unintuitive
> 
> Is this something that is forced by our choice of lexer/parser?
> 
> Historical note: originally, blackbird was meant to be kept as close to "python" as possible, since the way a user declares their program in SF is essentially also "blackbird code" (with some small, unavoidable, modifications between "standalone blackbird" and "python-embedded blackbird"). Why would we want to do that here, where---at least to my naive eye---it doesn't seem necessary to depart from python style?

@co9olguy There's no particular reason for choosing this style at all. I just played around with different semantics during the hackathon, but I could easily change it to be more python-like if that's preferred. :slightly_smiling_face: 

I chose to go with this since I thought it looked nice, and vaguely pythonic, while still following the style that I thought Blackbird had (e.g. few `:` and brackets/parentheses). Same with the array declarations, that were already there but that I simply expanded to include parameters (e.g. `{phase_0}`) and indexing. The code in [my last comment](https://github.com/XanaduAI/blackbird/pull/24#issuecomment-655792493) above shows better what I envisioned, but I have no issues with changing it to something else, or simply just making it look like python instead.

josh146 commented:
I'm in favour of making the syntax more Pythonic, with the caveat that Python doesn't actually have a native [`for` loop with syntactic counter](https://en.wikipedia.org/wiki/For_loop); instead, the `for` construct in Python is really a combination of a [`foreach` loop and generator functions like `range()`](https://en.wikipedia.org/wiki/Foreach_loop).

Implementing a native `for` loop would be much simpler than always requiring a `foreach` loop and iterator object.

We could:

* Implement separate `for` and `foreach` keyword
  ```python
  foreach int m in [0, 3, 1]
      MZgate(phases[m], phases[m+1]) | [m, m+1]

  for int m = 0,10,2
      MeasureFock() | m
  ```

* Use the same keyword for `for` and `foreach`, but distinguish the two based on another syntactic pattern:
  ```python
  for int m in [0, 3, 1]
      MZgate(phases[m], phases[m+1]) | [m, m+1]

  for int m = 0:10:2  # could also use 'in' here?
      MeasureFock() | m
  ```
  (e.g., in this case the syntax defining the loop counter `m` distinguishes the behaviour)

thisac commented:
[ch260]

thisac commented:
@josh146 @co9olguy I've changed the syntax for the for-loops to be more python-like. E.g.

```python
for int m in [2, 7, 3]
    MeasureFock() | m
```
to apply the operation to modes 2, 7 and 3, or

```python
par array phases =
    {phase_0}
    {phase_1}
    {phase_2}
    {phase_3}
    {phase_4}

for int m in 0:5:1
    MZgate(phases[m], phases[m+1]) | [m, m+1]
```
to apply the MZgate with indexing from the phases list for the range 0:5:1 = [0, 1, 2, 3, 4].

I also added some tests and fixed some issues that I found along the way. :slightly_smiling_face:

josh146 commented on blackbird_python/blackbird/listener.py:
```diff
@@ -66,6 +66,7 @@
     "int": int,
     "str": str,
     "bool": bool,
+    "par": sym.Symbol,
```
Where is the `par` datatype now used?

josh146 commented on blackbird_python/blackbird/listener.py:
```diff
@@ -135,7 +137,7 @@ def __init__(self, cwd=None):
         self._includes = {}
         self._cwd = cwd
-        self._infor = False
+        self._in_for = False
```
Much better 😆

josh146 commented on blackbird_python/blackbird/listener.py:
```diff
@@ -352,19 +354,15 @@ def exitStatement(self, ctx: blackbirdParser.StatementContext):
-                        raise ValueError(f"Mode must be of type int, not {type(_VAR[m])}")
-                else:
-                    raise ValueError(f"Variable {m} not declared")
+            m = _expression(m)
+
+            if isinstance(m, (int, np.integer)):
+                modes[i] = m
+            else:
+                raise ValueError("Mode must be of type int, not {}".format(type(m)))
```
Nice, added features while also simplifying the code!

josh146 commented on blackbird_python/blackbird/listener.py:
```diff
@@ -447,26 +445,39 @@ def exitStatement(self, ctx: blackbirdParser.StatementContext):
 
-        if ctx.INT():
-            _FORVAR = range(int(ctx.INT().getText()))
+        if ctx.rangeval():
```
Just double checking, but the range syntax can only be used inside a for loop. If it is used elsewhere, it will either (a) raise a syntax error (because the lexer won't be expecting it), or (b) if it passes the lexer, will be ignored by the parser since `if ctx.rangeval()` isn't included anywhere else.

thisac replied:
That's correct! It would be possible to add it as an allowed expression and thus be able to store it to an array, but I'm not sure that's worth it at the moment. I can't think of a clear use-case for it really. :thinking:

josh146 commented on blackbird_python/blackbird/listener.py:
```diff
@@ -447,26 +445,39 @@ def exitStatement(self, ctx: blackbirdParser.StatementContext):
-        if ctx.INT():
-            _FORVAR = range(int(ctx.INT().getText()))
+        if ctx.rangeval():
+            for_var = range(*[
+                int(c.getText()) for c in ctx.rangeval().getChildren()
+                if c.getText() != ":"
+            ])
```
This is quite nice, since I suppose
```python
for i in 5:10
```
and 
```python
for i in 5:10:2
```
will both work equally well, since `range()` only requires 2 arguments, but can accept 3.

thisac replied:
Yep, that would be the case. :slightly_smiling_face:

josh146 commented on blackbird_python/blackbird/listener.py:
```diff
@@ -447,26 +445,39 @@ def exitStatement(self, ctx: blackbirdParser.StatementContext):
-                    var = PYTHON_TYPES[ctx.vartype().getText()](var)
+                    new_var = PYTHON_TYPES[ctx.vartype().getText()](var)
+                    if new_var != var:
+                        raise ValueError
```
How come this `ValueError` doesn't contain a message?

I have to admit, I'm having a bit of trouble following this `try`-`except`. It could benefit from line comments!

thisac replied:
This `ValueError` will be excepted by the line below, thus never actually printed to the user. There might be a nicer way to do this (it's a bit hacky I guess). :thinking: 

So, basically, it tests to see if the variable is able to be casted to the defined type and, if so, it checks whether the casted variable has the same value as the old one (e.g. `5 != int(5.0)`) and if not, it raises the `invalid value` `ValueError`)

josh146 commented on blackbird_python/blackbird/listener.py:
```diff
@@ -447,26 +445,39 @@ def exitStatement(self, ctx: blackbirdParser.StatementContext):
-        _FORVAR = []
+
+        self._program._forvar[ctx.NAME().getText()] = np.array(for_var)
+        if ctx.NAME():
+            del _VAR[ctx.NAME().getText()]
```
This is to delete the temp for loop variable?

thisac replied:
Yes, otherwise the last element of the for-loop list will be left as a "free" variable, which isn't a huge problem but it's nicer if it's confined to the for-loop. It could cause issues later.

josh146 commented on blackbird_python/blackbird/tests/test_listener.py:
```diff
@@ -317,7 +302,7 @@ def test_operation_args_and_kwarg(self, parse_input_mocked_metadata):
         ]
 
-    @pytest.mark.parametrize("dc", [[1, 3], [1], 0])
+    @pytest.mark.parametrize("dc", [[1, 3], [1], 0, [True, 0]])
```
Nice!

josh146 commented on blackbird_python/blackbird/tests/test_listener.py:
```diff
@@ -344,6 +329,40 @@ def test_operation_arg_expressions(self, parse_input_mocked_metadata):
+        bb = parse_input_mocked_metadata(
+            "int m = 4\nMZgate(0, 1) | [m*2, -1+m]"
+        )
+        assert bb.operations == [{'op': 'MZgate', 'args': [0, 1], 'kwargs': {}, 'modes': [8, 3]}]
```
🎉

josh146 commented on blackbird_python/blackbird/tests/test_listener.py:
```diff
@@ -344,6 +329,40 @@ def test_operation_arg_expressions(self, parse_input_mocked_metadata):
+            {'op': 'MZgate', 'args': [sym.Symbol("phase_0"), sym.Symbol("phase_1")], 'kwargs': {}, 'modes': [0, 1]}
+        ]
+
+    def test_var_idx_in_modes(self, parse_input_mocked_metadata):
+        """Test that array indexing works inside modes"""
+        bb = parse_input_mocked_metadata(
+            "int array vars =\n\t1, 2\n\nMZgate(0, 1) | [vars[0], vars[1]]"
+        )
+        assert bb.operations == [
+            {'op': 'MZgate', 'args': [0, 1], 'kwargs': {}, 'modes': [1, 2]}
+        ]
```
This test tests the case
```python
int array vars =
    1, 2
```

Is it worth also testing the case
```python
int array vars =
    1
    2
```
(does this affect how you index?)

What about something like

```python
int array vars =
    1, 2
    3, 4
```

thisac replied:
I deliberately made the arrays always behave as flat, in the sense that you can write them (and store them) as nested numpy-arrays but they will behave as flat with regards to indexing. So your examples above would basically behave the same only with the last one having two more elements.

thisac replied:
I thought it looked nice to be able to list parameters horizontally rather than vertically,

```python
par array phases =
    {phase_0}
    {phase_1}
    {phase_2}
    {phase_3}
    {phase_4}
```

and vice versa with regards to values.

```python
int array vars =
    1, 2, 3, 4, 5, 42, 6
```

josh146 replied:
:+1:

josh146 commented on blackbird_python/blackbird/tests/test_listener.py:
```diff
@@ -372,6 +391,12 @@ def test_regref_transform(self, parse_input_mocked_metadata):
+    @pytest.mark.parametrize("m", [[3.5], ["4.2"], [1+2j], [True]])
+    def test_non_integer_mode(self, m, parse_input_mocked_metadata):
+        """Test that an error is used when setting non integer modes"""
+        with pytest.raises(ValueError, match="Mode must be of type int, not"):
```
:+1:

josh146 commented on blackbird_python/blackbird/tests/test_listener.py:
```diff
@@ -789,3 +814,42 @@ def test_parse_string(self):
+        with pytest.raises(ValueError, match="invalid value"):
+            bb = parse_input_mocked_metadata(
+                "for int m in [1, 4.2, 9]\n\tMZgate(0, 1) | [0, 1]"
+            )
```
Great tests! Might also be worth testing

* Indexing different shaped arrays (e.g., column vectors, matrices)

* The case where stepsize is ignored, if this is allowed (wasn't sure based on the tests above)

* Having expressions in the for loop syntax?

thisac replied:
Added tests for everything you mentioned. Realised that 

```python
int array vars =
    1, 2
    3, 4
```

works, while

```python
int array vars =
    1, 2
    3
```

does not; but I think this is a minor issue and can be looked into closer if/when we decide to rework the BB array type to e.g. include 2d indexes.

josh146 replied:
I think that should actually be fine, I'm not sure

```python
int array vars =
    1, 2
    3
```
is intended to/should be allowed. Since this corresponds to a 'ragged' numpy array

josh146 commented:
Something that just crossed my mind --- we will need to update the documentation, and especially the `syntax.rst` page, to take into account these changes!

In particular, to ensure that any deviations from Python are clearly laid out, such as flat indexing and for loops.
